crates/sui-graphql-rpc/src/types/transaction_block_effects.rs

```diff
@@ -106,40 +151,115 @@ impl TransactionBlockEffects {
     }
 
     /// Shared objects that are referenced by but not changed by this transaction.
-    async fn unchanged_shared_objects(&self) -> Option<Vec<UnchangedSharedObject>> {
-        Some(
-            self.native
-                .input_shared_objects()
-                .into_iter()
-                .filter_map(|input| UnchangedSharedObject::try_from(input).ok())
-                .collect(),
-        )
+    async fn unchanged_shared_objects(
+        &self,
+        ctx: &Context<'_>,
+        first: Option<u64>,
+        after: Option<CUnchangedSharedObject>,
+        last: Option<u64>,
+        before: Option<CUnchangedSharedObject>,
+    ) -> Result<Connection<String, UnchangedSharedObject>> {
+        let page = Page::from_params(ctx.data_unchecked(), first, after, last, before)?;
+        let mut connection = Connection::new(false, false);
+
+        // TODO: this iterates through its shared_objects field and collects.
+        // We can consider doing that collection here so we don't double dip.
+        let input_shared_objects = self.native.input_shared_objects();
+
+        let Some((prev, next, cs)) = page.paginate_indices(input_shared_objects.len()) else {
+            return Ok(connection);
+        };
+
+        connection.has_previous_page = prev;
+        connection.has_next_page = next;
+
+        for c in cs {
+            let result = UnchangedSharedObject::try_from(input_shared_objects[*c].clone());
+            match result {
+                Ok(unchanged_shared_object) => {
+                    connection
+                        .edges
+                        .push(Edge::new(c.encode_cursor(), unchanged_shared_object));
+                }
+                Err(_) => continue, // Drop failures
```

Comment:
I don't think we should, because this is an internal error, right? Let's just fast-fail.

Reply:
I see, the existing code has try_from().ok() https://github.com/MystenLabs/sui/blob/main/crates/sui-graphql-rpc/src/types/transaction_block_effects.rs#L114 but happy to make this an internal error instead

Reply:
Hm... when I do return an error on Input::Mutate, it looks like the test you had written at sui-graphql-e2e-tests/tests/transactions/shared.move fails. It looks like you had touched this last so I'm sure you have the most context, but I think this is an "error" in the sense that some shared object was of variant Mutate, so we return an error since we're looking for UnchangedSharedObject. in the resolver code though, this is not actually an error, we just drop it - since the shared object was mutated, it wouldn't make sense to render it, but it also wouldn't make sense to error on it.

Reply:
Ah okay, then yes, I agree with you -- (and me in the past, I suppose 😅) we should `continue` on errors.